pkg/types/container/v0.0.1/entry.go

```diff
@@ -0,0 +1,329 @@
+	if err := json.Unmarshal(artifactBytes, img); err != nil {
+		return err
+	}
+	v.ContainerObj.Data.Content = img
+	v.ContainerObj.Data.Hash = &models.ContainerV001SchemaDataHash{}
+	v.ContainerObj.Data.Hash.Algorithm = swag.String(models.ContainerV001SchemaDataHashAlgorithmSha256)
+	h := sha256.Sum256(artifactBytes)
+	v.ContainerObj.Data.Hash.Value = swag.String(hex.EncodeToString(h[:]))
+
+	return nil
+}
+
+func (v *V001Entry) Canonicalize(ctx context.Context) ([]byte, error) {
+	// At this point SHA should have already been populated.
+	keyObj, sigObj, err := v.validate()
+	if err != nil {
+		return nil, err
+	}
+
+	canonicalEntry := models.ContainerV001Schema{}
+
+	// need to canonicalize signature & key content
+	canonicalEntry.Signature = &models.ContainerV001SchemaSignature{}
+	// signature URL (if known) is not set deliberately
+	canonicalEntry.Signature.Content, err = sigObj.CanonicalValue()
+	if err != nil {
+		return nil, err
+	}
+
+	// key URL (if known) is not set deliberately
+	canonicalEntry.Signature.PublicKey = &models.ContainerV001SchemaSignaturePublicKey{}
+	canonicalEntry.Signature.PublicKey.Content, err = keyObj.CanonicalValue()
+	if err != nil {
+		return nil, err
+	}
+
+	// both content (for image reference) and hash (for verification) are specified
+	canonicalEntry.Data = v.ContainerObj.Data
+
+	// wrap in valid object with kind and apiVersion set
+	containerObj := models.Container{}
+	containerObj.APIVersion = swag.String(APIVERSION)
+	containerObj.Spec = &canonicalEntry
+
+	v.ContainerObj = canonicalEntry
+
+	bytes, err := json.Marshal(&containerObj)
+	if err != nil {
+		return nil, err
+	}
+
+	return bytes, nil
+}
+
+// validate performs cross-field validation for fields in object
+func (v V001Entry) validate() (pki.PublicKey, pki.Signature, error) {
+	sig := v.ContainerObj.Signature
+	if v.ContainerObj.Signature == nil {
+		return nil, nil, types.ValidationError(errors.New("missing signature"))
+	}
+	if len(sig.Content) == 0 {
+		return nil, nil, errors.New("'content' must be specified for signature")
+	}
+	sigObj, err := x509.NewSignature(bytes.NewReader(sig.Content))
+	if err != nil {
+		return nil, nil, types.ValidationError(err)
+	}
+
+	key := sig.PublicKey
+	if key == nil {
+		return nil, nil, types.ValidationError(errors.New("missing public key"))
+	}
+	if len(key.Content) == 0 {
+		return nil, nil, errors.New("'content'  must be specified for publicKey")
+	}
+	keyObj, err := x509.NewPublicKey(bytes.NewReader(key.Content))
+	if err != nil {
+		return nil, nil, types.ValidationError(err)
+	}
+
+	data := v.ContainerObj.Data
+	if data == nil {
+		return nil, nil, errors.New("missing data")
+	}
+	if data.Content == nil {
+		return nil, nil, errors.New("'content' must be specified for data")
+	}
+
+	hash := data.Hash
+	if hash == nil {
+		return nil, nil, errors.New("'hash' must be specified for data")
+	}
+	decoded, err := hex.DecodeString(*hash.Value)
+	if err != nil {
+		return nil, nil, err
+	}
```

Comment:
On second thought I don't think we need it at all -- originally it was there for fetching the content by a URL ref, and then I thought to keep it in case you didn't want to upload the signed content. but at that point just use hashedrekord.

Reply:
Hmmm @bobcallaway I was just running the tests and finishing up addressing these comments but I have a major problem:

without a canonicalized hash of the content, there's no good way of verifying the entry. someone could send proposed entry with a a valid sig over a sha256 hash and then put whatever JSON payload they want. so entry verification really has to match the JSON payload with the sha256hash. 

i feel like there should be a way of canonicalization. but since signing w/ cosign is a signature over the payload bytes of the signed manifest, maybe I should stick the marshalled artifact bytes into the entry, rather than the unmarshalled interface